Stats reporter don't show biggest allocation in the histogram

Consider this code:

```
import mmap

with mmap.mmap(-1, 10000000000):
    pass
```

This is what we currently show in the `stat` reporter:

```
📏 Total allocations:
        20

📦 Total memory allocated:
        9.313GB

📊 Histogram of allocation size:
        min: 528.000B
        ----------------------------------------
        < 2.754KB  : 8 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
        < 14.716KB : 2 ▇▇▇▇▇▇▇
        < 78.621KB : 0
        < 420.029KB: 0
        < 2.191MB  : 0
        < 11.707MB : 0
        < 62.545MB : 0
        < 334.139MB: 0
        < 1.743GB  : 0
        <=9.313GB  : 0
        ----------------------------------------
        max: 9.313GB

📂 Allocator type distribution:
         MALLOC: 10
         MMAP: 1
```

The `mmap` allocation is not shown in the histogram (is the only one missing).

These are the failing tests 

```

memray on  main [?⇡] via  v10.19.0 via 🐍 v3.8.10 (memray-env)
🕙21:42:31 [🧱 INT] ❯ pytest --memray tests/unit
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.8.10, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/jingkai/Documents/github/memray/memray, configfile: pyproject.toml
plugins: cov-3.0.0, memray-1.0.0
collected 187 items

tests/unit/test_cli.py ...........................................................                                                                                                                            [ 31%]
tests/unit/test_flamegraph_reporter.py ....F.......                                                                                                                                                           [ 37%]
tests/unit/test_frame_tools.py ............                                                                                                                                                                   [ 44%]
tests/unit/test_highwatermark_command.py ............                                                                                                                                                         [ 50%]
tests/unit/test_reader.py FF.F                                                                                                                                                                                [ 52%]
tests/unit/test_stats_reporter.py ......FF.FF.                                                                                                                                                                [ 59%]
tests/unit/test_summary_reporter.py ....                                                                                                                                                                      [ 61%]
tests/unit/test_table_reporter.py .....                                                                                                                                                                       [ 64%]
tests/unit/test_templates.py ....                                                                                                                                                                             [ 66%]
tests/unit/test_tracker.py FF                                                                                                                                                                                 [ 67%]
tests/unit/test_tree_reporter.py ..........................                                                                                                                                                   [ 81%]
tests/unit/test_tui_reporter.py ...................................                                                                                                                                           [100%]

===================================================================================================== FAILURES ======================================================================================================
__________________________________________________________________________ TestFlameGraphReporter.test_sanity_check_with_real_allocations ___________________________________________________________________________

self = <tests.unit.test_flamegraph_reporter.TestFlameGraphReporter object at 0x7f03056b7ee0>, tmp_path = PosixPath('/tmp/pytest-of-jingkai/pytest-1/test_sanity_check_with_real_al0')

    def test_sanity_check_with_real_allocations(self, tmp_path):
        # GIVEN
        allocator = MemoryAllocator()
        output = tmp_path / "test.bin"
>       with Tracker(output):

tests/unit/test_flamegraph_reporter.py:243:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError
________________________________________________________________________________________ test_rejects_different_header_magic ________________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-jingkai/pytest-1/test_rejects_different_header_0')

    def test_rejects_different_header_magic(tmp_path):
        # GIVEN
        output = tmp_path / "test.bin"
        allocator = MemoryAllocator()

        # WHEN
        # Create a valid allocation record file
>       with Tracker(output):

tests/unit/test_reader.py:18:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError
_______________________________________________________________________________________ test_rejects_different_header_version _______________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-jingkai/pytest-1/test_rejects_different_header_1')

    def test_rejects_different_header_version(tmp_path):
        # GIVEN
        output = tmp_path / "test.bin"
        allocator = MemoryAllocator()

        # WHEN
        # Create a valid allocation record file
        destination = FileDestination(output, overwrite=False, compress_on_exit=False)
>       with Tracker(destination=destination):

tests/unit/test_reader.py:40:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError
___________________________________________________________________________________________________ test_read_pid ___________________________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-jingkai/pytest-1/test_read_pid0')

    def test_read_pid(tmp_path):
        # GIVEN
        output = tmp_path / "test.bin"
        allocator = MemoryAllocator()

        # WHEN
>       with Tracker(output):

tests/unit/test_reader.py:77:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError
____________________________________________________________________________________________ test_get_histogram_databins ____________________________________________________________________________________________

    def test_get_histogram_databins():
        # GIVEN
        input_data = [
            2500,
            11000,
            11000,
            12000,
            60000,
            65000,
            120000,
            125000,
            125000,
            160000,
            170000,
            180000,
            800000,
            1500000,
        ]
        expected_output = [
            (8986, 1),
            (32299, 3),
            (116099, 2),
            (417312, 6),
            (1500000, 1),
        ]

        # WHEN
        actual_output = get_histogram_databins(input_data, bins=5)

        # THEN
>       assert expected_output == actual_output
E       assert [(8986, 1), (... (1500000, 1)] == [(8986, 1), (... (5391647, 1)]
E         Right contains one more item: (5391647, 1)
E         Use -v to get more diff

tests/unit/test_stats_reporter.py:242: AssertionError
_______________________________________________________________________________________ test_get_histogram_databins_rounding ________________________________________________________________________________________

    def test_get_histogram_databins_rounding():
        """Data chosen to provoke a floating point rounding error.

        In particular, so that:

            log(low) + sum([(log(high) - log(low)) / bins] * bins) > log(high)
        """
        # GIVEN
        input_data = [
            32,
            1050856,
        ]
        expected_output = [
            (90, 1),
            (256, 0),
            (724, 0),
            (2049, 0),
            (5798, 0),
            (16405, 0),
            (46411, 0),
            (131299, 0),
            (371453, 0),
            (1050856, 1),
        ]

        # WHEN
        actual_output = get_histogram_databins(input_data, bins=10)

        # THEN
>       assert expected_output == actual_output
E       assert [(90, 1), (25...6405, 0), ...] == [(90, 1), (25...6405, 0), ...]
E         Right contains one more item: (2972915, 0)
E         Use -v to get more diff

tests/unit/test_stats_reporter.py:274: AssertionError
________________________________________________________________________________________________ test_draw_histogram ________________________________________________________________________________________________

    def test_draw_histogram():
        # GIVEN
        input_data = [
            2500,
            11000,
            11000,
            12000,
            60000,
            65000,
            120000,
            125000,
            125000,
            160000,
            170000,
            180000,
            800000,
            1500000,
        ]
        expected_output = """min: 2.441KB
    \t----------------------------------------
    \t< 8.775KB  : 1 ▇▇▇▇▇
    \t< 31.542KB : 3 ▇▇▇▇▇▇▇▇▇▇▇▇▇
    \t< 113.378KB: 2 ▇▇▇▇▇▇▇▇▇
    \t< 407.531KB: 6 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
    \t<=1.431MB  : 1 ▇▇▇▇▇
    \t----------------------------------------
    \tmax: 1.431MB"""

        # WHEN
        actual_output = draw_histogram(input_data, bins=5)

        # THEN
>       assert expected_output == actual_output
E       AssertionError: assert 'min: 2.441KB...tmax: 1.431MB' == 'min: 2.441KB...tmax: 1.431MB'
E         Skipping 167 identical leading characters in diff, use -v to show
E           ▇▇▇▇▇▇▇
E         -     < 1.431MB  : 1 ▇▇▇▇▇
E         ?      ^
E         +     <=1.431MB  : 1 ▇▇▇▇▇
E         ?      ^
E         -     <=5.142MB  : 1 ▇▇▇▇▇...
E
E         ...Full output truncated (3 lines hidden), use '-vv' to show

tests/unit/test_stats_reporter.py:316: AssertionError
_____________________________________________________________________________________ test_draw_histogram_smaller_scale_factor ______________________________________________________________________________________

    def test_draw_histogram_smaller_scale_factor():
        # GIVEN
        input_data = [
            2500,
            11000,
            11000,
            12000,
            60000,
            65000,
            120000,
            125000,
            125000,
            160000,
            170000,
            180000,
            800000,
            1500000,
        ]

        expected_output = """min: 2.441KB
    \t--------------------
    \t< 8.775KB  : 1 ▇
    \t< 31.542KB : 3 ▇▇▇
    \t< 113.378KB: 2 ▇▇
    \t< 407.531KB: 6 ▇▇▇▇▇
    \t<=1.431MB  : 1 ▇
    \t--------------------
    \tmax: 1.431MB"""

        # WHEN
        actual_output = draw_histogram(
            input_data, bins=5, hist_scale_factor=5
        )  # setting this to 5

        # THEN
>       assert expected_output == actual_output
E       AssertionError: assert 'min: 2.441KB...tmax: 1.431MB' == 'min: 2.441KB...tmax: 1.431MB'
E         Skipping 106 identical leading characters in diff, use -v to show
E           6 ▇▇▇▇▇
E         -     < 1.431MB  : 1 ▇
E         ?      ^
E         +     <=1.431MB  : 1 ▇
E         ?      ^
E         -     <=5.142MB  : 1 ▇...
E
E         ...Full output truncated (3 lines hidden), use '-vv' to show

tests/unit/test_stats_reporter.py:354: AssertionError
________________________________________________________________________________ test_two_trackers_cannot_be_active_at_the_same_time ________________________________________________________________________________

tmpdir = local('/tmp/pytest-of-jingkai/pytest-1/test_two_trackers_cannot_be_ac0')

    def test_two_trackers_cannot_be_active_at_the_same_time(tmpdir):
        # GIVEN
        output = Path(tmpdir) / "test.bin"
        output2 = Path(tmpdir) / "test2.bin"

        # WHEN
>       with Tracker(output):

tests/unit/test_tracker.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError
__________________________________________________________________________________ test_the_same_tracker_cannot_be_activated_twice __________________________________________________________________________________

tmpdir = local('/tmp/pytest-of-jingkai/pytest-1/test_the_same_tracker_cannot_b0')

    def test_the_same_tracker_cannot_be_activated_twice(tmpdir):
        # GIVEN
        output = Path(tmpdir) / "test.bin"

        # WHEN
        tracker = Tracker(output)
>       with tracker:

tests/unit/test_tracker.py:27:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise RuntimeError("No more than one Tracker instance can be active at the same time")
E   RuntimeError: No more than one Tracker instance can be active at the same time

src/memray/_memray.pyx:305: RuntimeError

=================================================================================================== MEMRAY REPORT ===================================================================================================
Allocations results for tests/unit/test_tree_reporter.py::TestTreeReporter::test_very_deep_call_is_limited

         📦 Total memory allocated: 531.9KiB
         📏 Total allocations: 112
         📊 Histogram of allocation sizes: |█ █ █    |
         🥇 Biggest allocating functions:
                - collapse_tree:/home/jingkai/Documents/github/memray/memray/src/memray/reporters/tree.py:41 -> 256.0KiB
                - __init__:<string>:5 -> 256.0KiB
                - <listcomp>:/home/jingkai/Documents/github/memray/memray/tests/unit/test_tree_reporter.py:1035 -> 16.1KiB
                - collapse_tree:/home/jingkai/Documents/github/memray/memray/src/memray/reporters/tree.py:42 -> 3.2KiB
                - wrapper:/home/jingkai/Documents/github/memray/memray-env/lib/python3.8/site-packages/pytest_memray/plugin.py:98 -> 656.0B


Allocations results for tests/unit/test_cli.py::TestRunSubCommand::test_run_with_live

         📦 Total memory allocated: 353.5KiB
         📏 Total allocations: 1572
         📊 Histogram of allocation sizes: |█ ▇ ▁    |
         🥇 Biggest allocating functions:
                - __init__:/usr/lib/python3.8/unittest/mock.py:2076 -> 256.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2030 -> 12.9KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2026 -> 10.0KiB
                - __setattr__:/usr/lib/python3.8/unittest/mock.py:740 -> 6.5KiB
                - __new__:/usr/lib/python3.8/unittest/mock.py:418 -> 5.5KiB


Allocations results for tests/unit/test_cli.py::TestRunSubCommand::test_run_module

         📦 Total memory allocated: 336.2KiB
         📏 Total allocations: 1558
         📊 Histogram of allocation sizes: |█ ▅      |
         🥇 Biggest allocating functions:
                - __init__:/usr/lib/python3.8/unittest/mock.py:2076 -> 256.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2026 -> 10.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2030 -> 8.6KiB
                - register:/usr/lib/python3.8/argparse.py:1322 -> 4.5KiB
                - __setattr__:/usr/lib/python3.8/unittest/mock.py:740 -> 4.3KiB


Allocations results for tests/unit/test_cli.py::TestRunSubCommand::test_run_default_output

         📦 Total memory allocated: 325.8KiB
         📏 Total allocations: 1355
         📊 Histogram of allocation sizes: |█ ▅      |
         🥇 Biggest allocating functions:
                - __init__:/usr/lib/python3.8/argparse.py:215 -> 256.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2026 -> 10.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2030 -> 8.6KiB
                - register:/usr/lib/python3.8/argparse.py:1322 -> 4.5KiB
                - __setattr__:/usr/lib/python3.8/unittest/mock.py:740 -> 4.3KiB


Allocations results for tests/unit/test_cli.py::TestRunSubCommand::test_run_with_native_mode

         📦 Total memory allocated: 324.7KiB
         📏 Total allocations: 1382
         📊 Histogram of allocation sizes: |█ ▆ ▁    |
         🥇 Biggest allocating functions:
                - decode:/usr/lib/python3.8/os.py:753 -> 256.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2026 -> 10.0KiB
                - _mock_set_magics:/usr/lib/python3.8/unittest/mock.py:2030 -> 8.6KiB
                - register:/usr/lib/python3.8/argparse.py:1322 -> 4.5KiB
                - __setattr__:/usr/lib/python3.8/unittest/mock.py:740 -> 4.3KiB


============================================================================================== short test summary info ==============================================================================================
FAILED tests/unit/test_flamegraph_reporter.py::TestFlameGraphReporter::test_sanity_check_with_real_allocations - RuntimeError: No more than one Tracker instance can be active at the same time
FAILED tests/unit/test_reader.py::test_rejects_different_header_magic - RuntimeError: No more than one Tracker instance can be active at the same time
FAILED tests/unit/test_reader.py::test_rejects_different_header_version - RuntimeError: No more than one Tracker instance can be active at the same time
FAILED tests/unit/test_reader.py::test_read_pid - RuntimeError: No more than one Tracker instance can be active at the same time
FAILED tests/unit/test_stats_reporter.py::test_get_histogram_databins - assert [(8986, 1), (... (1500000, 1)] == [(8986, 1), (... (5391647, 1)]
FAILED tests/unit/test_stats_reporter.py::test_get_histogram_databins_rounding - assert [(90, 1), (25...6405, 0), ...] == [(90, 1), (25...6405, 0), ...]
FAILED tests/unit/test_stats_reporter.py::test_draw_histogram - AssertionError: assert 'min: 2.441KB...tmax: 1.431MB' == 'min: 2.441KB...tmax: 1.431MB'
FAILED tests/unit/test_stats_reporter.py::test_draw_histogram_smaller_scale_factor - AssertionError: assert 'min: 2.441KB...tmax: 1.431MB' == 'min: 2.441KB...tmax: 1.431MB'
FAILED tests/unit/test_tracker.py::test_two_trackers_cannot_be_active_at_the_same_time - RuntimeError: No more than one Tracker instance can be active at the same time
FAILED tests/unit/test_tracker.py::test_the_same_tracker_cannot_be_activated_twice - RuntimeError: No more than one Tracker instance can be active at the same time
========================================================================================== 10 failed, 177 passed in 1.52s ===========================================================================================
```

--- a/news/95.bugfix.rst
+++ b/news/95.bugfix.rst
@@ -1,0 +1,1 @@
+Fixed a bug in the :doc:`stats reporter <stats>` that could result in the largest allocations being omitted from the histogram.

--- a/src/memray/reporters/stats.py
+++ b/src/memray/reporters/stats.py
@@ -91,7 +91,7 @@ def get_histogram_databins(data: List[int], bins: int) -> List[Tuple[int, int]]:
 
     # Determine the upper bound in bytes for each bin
     steps = [int(math.exp(low + step * (i + 1))) for i in range(bins)]
-    dist = Counter((x - low) // step for x in it)
+    dist = Counter(min((x - low) // step, bins - 1) for x in it)
     return [(steps[b], dist[b]) for b in range(bins)]
 
 
@@ -112,7 +112,6 @@ def draw_histogram(data: List[int], bins: int, *, hist_scale_factor: int = 25) -
         )
 
     data_bins = get_histogram_databins(data, bins=bins)
-
     max_data_bin = max([t[1] for t in data_bins])
     scaled_data_bins = [
         math.ceil((v / max_data_bin) * hist_scale_factor) for _, v in data_bins

--- a/tests/unit/test_stats_reporter.py
+++ b/tests/unit/test_stats_reporter.py
@@ -232,7 +232,7 @@ def test_get_histogram_databins():
         (32299, 3),
         (116099, 2),
         (417312, 6),
-        (1500000, 1),
+        (1500000, 2),
     ]
 
     # WHEN
@@ -274,6 +274,34 @@ def test_get_histogram_databins_rounding():
     assert expected_output == actual_output
 
 
+def test_get_histogram_over_bound():
+    """Data chosen to provoke a scenario where the computed allocation exceeds the upper limit.
+
+    In particular, so that:
+        Counter(min((x - low) // step, bins-1) for x in it) will default to placing it in the
+        last bin instead of creating a new record out of range of the bins.
+    """
+    input_data = [10000000000, 536, 536, 592, 576, 4486]
+    expected_output = [
+        (2859, 4),
+        (15252, 1),
+        (81360, 0),
+        (434009, 0),
+        (2315167, 0),
+        (12349970, 0),
+        (65879369, 0),
+        (351425246, 0),
+        (1874633954, 0),
+        (10000000000, 1),
+    ]
+
+    # WHEN
+    actual_output = get_histogram_databins(input_data, bins=10)
+
+    # THEN
+    assert expected_output == actual_output
+
+
 def test_get_histogram_databins_invalid_bins():
     with pytest.raises(ValueError):
         _ = get_histogram_databins([], bins=0)  # invalid bins value
@@ -305,7 +333,7 @@ def test_draw_histogram():
 \t< 31.542KB : 3 ▇▇▇▇▇▇▇▇▇▇▇▇▇
 \t< 113.378KB: 2 ▇▇▇▇▇▇▇▇▇
 \t< 407.531KB: 6 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
-\t<=1.431MB  : 1 ▇▇▇▇▇
+\t<=1.431MB  : 2 ▇▇▇▇▇▇▇▇▇
 \t----------------------------------------
 \tmax: 1.431MB"""
 
@@ -341,7 +369,7 @@ def test_draw_histogram_smaller_scale_factor():
 \t< 31.542KB : 3 ▇▇▇
 \t< 113.378KB: 2 ▇▇
 \t< 407.531KB: 6 ▇▇▇▇▇
-\t<=1.431MB  : 1 ▇
+\t<=1.431MB  : 2 ▇▇
 \t--------------------
 \tmax: 1.431MB"""
 

